`jets deploy` fails on a newly created project which uses VPC

<!--
Hi! Thanks for considering to file a bug with Jets. Please take the time to
answer the basic questions. Please try to be as detailed as possible.
To be sensitive to everyone's time, if not enough details are provided, the
issue may be closed without comment. If you repeatedly fail to provide enough
details, you may be blocked from ever submitting issues to Jets again.
Please use your best judgment. 👍

If you are unsure this is a bug in Jets, please consider asking your question at:
https://community.boltops.com

Thanks!
-->

## Checklist

<!--
Make sure that you've done all of these.  To mark a checkbox done, replace [ ] with [x]. Or after you create the issue you can click the checkbox.
-->

- [x] Upgrade Jets: Are you using the latest version of Jets? This allows Jets to fix issues fast. There's a `jets upgrade` command that makes this a simple task. There's also an Upgrading Guide: http://rubyonjets.com/docs/upgrading/
- [x] Reproducibility: Are you reporting a bug others will be able to reproduce and not asking a question. If you're unsure or want to ask a question, do so on https://community.boltops.com
- [ ] Code sample: Have you put together a code sample to reproduce the issue and make it available?  Code samples help speed up fixes dramatically. If it's an easily reproducible issue, then code samples are not needed. If you're unsure, please include a code sample.

## My Environment

<!-- Please fill out the table below with debugging info to help: -->

| Software         | Version |
| ---------------- | ------- |
| Operating System | Amazon Linux 2 |
| Jets             | 4.0.3 |
| Ruby             | 3.2.2 |

---

## Expected Behaviour

<!--
What is it you expected to happen? This should be a description of how the functionality you tried to use is supposed to work. Try to keep this to one-paragraph.
-->
`jets deploy` to success.

## Current Behavior

<!--
Describe the details of the bug.  Try to keep this to one-paragraph.
-->
`jets deploy` fails with the error:
```
12:11:24AM UPDATE_IN_PROGRESS AWS::CloudFormation::Stack test-jets-api-dev User Initiated
12:11:27AM CREATE_IN_PROGRESS AWS::CloudFormation::Stack ApiGateway 
12:11:27AM CREATE_IN_PROGRESS AWS::Lambda::LayerVersion GemLayer 
12:11:27AM CREATE_IN_PROGRESS AWS::IAM::Role IamRole 
12:11:28AM CREATE_IN_PROGRESS AWS::IAM::Role IamRole Resource creation Initiated
12:11:28AM CREATE_IN_PROGRESS AWS::CloudFormation::Stack ApiGateway Resource creation Initiated
12:11:37AM CREATE_IN_PROGRESS AWS::Lambda::LayerVersion GemLayer Resource creation Initiated
12:11:37AM CREATE_COMPLETE AWS::Lambda::LayerVersion GemLayer 
12:11:38AM CREATE_COMPLETE AWS::CloudFormation::Stack ApiGateway 
12:11:39AM CREATE_IN_PROGRESS AWS::CloudFormation::Stack ApiResources1 
12:11:40AM CREATE_IN_PROGRESS AWS::CloudFormation::Stack ApiResources1 Resource creation Initiated
12:11:43AM CREATE_COMPLETE AWS::IAM::Role IamRole 
12:11:44AM CREATE_IN_PROGRESS AWS::IAM::Policy IamPolicy 
12:11:44AM CREATE_IN_PROGRESS AWS::CloudFormation::Stack JetsPreheatJob 
12:11:45AM CREATE_IN_PROGRESS AWS::CloudFormation::Stack JetsPreheatJob Resource creation Initiated
12:11:46AM CREATE_IN_PROGRESS AWS::IAM::Policy IamPolicy Resource creation Initiated
12:11:50AM CREATE_COMPLETE AWS::CloudFormation::Stack ApiResources1 
12:11:51AM CREATE_IN_PROGRESS AWS::CloudFormation::Stack JetsPublicController 
12:11:52AM CREATE_IN_PROGRESS AWS::CloudFormation::Stack JetsPublicController Resource creation Initiated
12:12:02AM CREATE_COMPLETE AWS::IAM::Policy IamPolicy 
12:12:02AM CREATE_FAILED AWS::CloudFormation::Stack JetsPublicController Embedded stack arn:aws:cloudformation:ap-northeast-1:xxxxxxxxxxxx:stack/test-jets-api-dev-JetsPublicController-xxxxxxxxxxxxx/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx was not successfully created: The following resource(s) failed to create: [ShowLambdaFunction]. 
12:12:02AM CREATE_FAILED AWS::CloudFormation::Stack JetsPreheatJob Resource creation cancelled
12:12:03AM UPDATE_ROLLBACK_IN_PROGRESS AWS::CloudFormation::Stack test-jets-api-dev The following resource(s) failed to create: [JetsPublicController, JetsPreheatJob]. 
12:12:06AM UPDATE_ROLLBACK_COMPLETE_CLEANUP_IN_PROGRESS AWS::CloudFormation::Stack test-jets-api-dev 
12:12:07AM DELETE_IN_PROGRESS AWS::IAM::Policy IamPolicy 
12:12:07AM DELETE_IN_PROGRESS AWS::CloudFormation::Stack JetsPreheatJob 
12:12:07AM DELETE_IN_PROGRESS AWS::CloudFormation::Stack JetsPublicController 
12:12:08AM DELETE_COMPLETE AWS::IAM::Policy IamPolicy 
12:12:18AM DELETE_COMPLETE AWS::CloudFormation::Stack JetsPublicController 
12:12:18AM DELETE_IN_PROGRESS AWS::CloudFormation::Stack ApiResources1 
12:12:29AM DELETE_COMPLETE AWS::CloudFormation::Stack JetsPreheatJob 
12:12:29AM DELETE_COMPLETE AWS::CloudFormation::Stack ApiResources1 
12:12:29AM DELETE_IN_PROGRESS AWS::IAM::Role IamRole 
12:12:30AM DELETE_IN_PROGRESS AWS::Lambda::LayerVersion GemLayer 
12:12:30AM DELETE_IN_PROGRESS AWS::CloudFormation::Stack ApiGateway 
12:12:31AM DELETE_COMPLETE AWS::Lambda::LayerVersion GemLayer 
12:12:31AM DELETE_COMPLETE AWS::IAM::Role IamRole 
12:12:40AM DELETE_COMPLETE AWS::CloudFormation::Stack ApiGateway 
12:12:41AM UPDATE_ROLLBACK_COMPLETE AWS::CloudFormation::Stack test-jets-api-dev 
Stack rolled back: UPDATE_ROLLBACK_COMPLETE
Time took: 1m 16s
The Jets application failed to deploy. Jets creates a few CloudFormation stacks to deploy your application.
The logs above show the CloudFormation parent stack events and points to the stack with the error.
Please go to the CloudFormation console and look for the specific stack with the error.
The specific child stack usually shows more detailed information and can be used to resolve the issue.
Example of checking the CloudFormation console: https://rubyonjets.com/docs/debugging/cloudformation/
```

## Step-by-step reproduction instructions

<!--
Be sure to include any steps you took for the problem to exist. This is likely the longest part of the report.

Please include any logs you think relevant here. If the logs are long (more than 50 lines) please make a gist of the logs and link to it. https://gist.github.com

With long logs, you can also use the <details> tag to keep the report readable. Example:

<details>
 <summary>Summary Goes Here</summary>

 ...this is hidden, collapsable content. start with a blank line to get terminal output to format right...
</details>
-->
* Create a new api project. `jets new test-jets-api --mode api`
* Configure the project to use VPC. (see: https://rubyonjets.com/docs/considerations/vpc/)
* Deploy the project. `AWS_REGION=ap-northeast-1 jets deploy`

## Code Sample

<!--
Please provide a code repository, gist, code snippet or sample files to reproduce the issue.
-->

## Solution Suggestion

<!--
Please provide possible solutions. If you can't think of anything, feel free to omit. Please be kind and add helpful possible solutions.  For example, "Fix it!" is not a helpful solution suggestion. We are mere mortals. Please be constructive.
-->
The cause of the error is that the role lacks CreateNetworkInterface permission:
"The provided execution role does not have permissions to call CreateNetworkInterface on EC2" (from CloudTrail event log of CreateFunction20150331)
This may be a timing issue.
The IAM Policy is not attached to the IAM Role when the IAM Role is created. So subsequent resource creations were failed.
(IAM Role and its policy was separated from this commit https://github.com/boltops-tools/jets/commit/8f960ef874c4f60e0d453f47e2100ddd9665f2fc)
I resolved this issue by adding `DependsOn: IamPolicy` for every resource that requires IAMRole to the CloudFormation files(parent stack and child stack).
But it also breaks Prewarming functionalities...

--- a/lib/jets/application/defaults.rb
+++ b/lib/jets/application/defaults.rb
@@ -2,8 +2,8 @@ class Jets::Application
   module Defaults
     extend ActiveSupport::Concern
 
-    included do
-      def self.default_iam_policy
+    class_methods do
+      def default_iam_policy
         project_namespace = Jets.project_namespace
         logs = {
           action: ["logs:*"],
@@ -24,23 +24,22 @@ class Jets::Application
         }
         policies << cloudformation
 
-        if Jets.config.function.vpc_config
-          vpc = {
-            action: %w[
-              ec2:CreateNetworkInterface
-              ec2:DeleteNetworkInterface
-              ec2:DescribeNetworkInterfaces
-              ec2:DescribeVpcs
-              ec2:DescribeSubnets
-              ec2:DescribeSecurityGroups
-            ],
-            effect: "Allow",
-            resource: "*",
-          }
-          policies << vpc
-        end
-
         policies
+      end
+
+      def vpc_iam_policy_statement
+        {
+          Action: %w[
+            ec2:CreateNetworkInterface
+            ec2:DeleteNetworkInterface
+            ec2:DescribeNetworkInterfaces
+            ec2:DescribeVpcs
+            ec2:DescribeSubnets
+            ec2:DescribeSecurityGroups
+          ],
+          Effect: "Allow",
+          Resource: "*",
+        }
       end
     end
 
@@ -200,6 +199,30 @@ class Jets::Application
         app/functions
         app/shared/functions
       ]
+    end
+
+    # Used by app/jobs/jets/preheat_job.rb
+    def preheat_job_iam_policy
+      policy = [
+        {
+          Sid: "Statement1",
+          Action: ["logs:*"],
+          Effect: "Allow",
+          Resource: [{
+            "Fn::Sub": "arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:/aws/lambda/${JetsPreheatJobWarmLambdaFunction}"
+          }]
+        },
+        {
+          Sid: "Statement2",
+          Action: ["lambda:InvokeFunction", "lambda:InvokeAsync"],
+          Effect: "Allow",
+          Resource: [{
+            "Fn::Sub": "arn:aws:lambda:${AWS::Region}:${AWS::AccountId}:function:#{Jets.project_namespace}-*"
+          }]
+        }
+      ]
+      policy << Jets::Application.vpc_iam_policy_statement if Jets.config.function.vpc_config
+      policy
     end
   end
 end

--- a/lib/jets/internal/app/jobs/jets/preheat_job.rb
+++ b/lib/jets/internal/app/jobs/jets/preheat_job.rb
@@ -7,24 +7,7 @@ class Jets::PreheatJob < ApplicationJob
 
   class_timeout 30
   class_memory 1024
-  class_iam_policy(
-    {
-      sid: "Statement1",
-      action: ["logs:*"],
-      effect: "Allow",
-      resource: [
-        sub("arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:/aws/lambda/${WarmLambdaFunction}"),
-      ]
-    },
-    {
-      Sid: "Statement2",
-      Action: ["lambda:InvokeFunction", "lambda:InvokeAsync"],
-      Effect: "Allow",
-      Resource: [
-        sub("arn:aws:lambda:${AWS::Region}:${AWS::AccountId}:function:#{Jets.project_namespace}-*")
-      ]
-    }
-  )
+  class_iam_policy(Jets.config.preheat_job_iam_policy)
 
   rate(PREWARM_RATE) if torching
   def torch

--- a/lib/jets/resource/iam/base_role_definition.rb
+++ b/lib/jets/resource/iam/base_role_definition.rb
@@ -24,11 +24,25 @@ module Jets::Resource::Iam
         }
       }
 
+      # Add vpc permissions to all policies
+      definition[logical_id][:properties][:policies] = [
+        policy_name: "vpc", # required, limited to 128-chars
+        policy_document: vpc_policy_document,
+      ] if vpc_policy_document
+
       unless managed_policy_arns.empty?
         definition[logical_id][:properties][:managed_policy_arns] = managed_policy_arns
       end
 
       definition
+    end
+
+    def vpc_policy_document
+      if Jets.config.function.vpc_config
+        {
+          Statement: [Jets::Application.vpc_iam_policy_statement]
+        }
+      end
     end
 
     def policy_document

